site/content/posts/2021-06-21-windows-e2e.md

```diff
@@ -0,0 +1,134 @@
+
+But running just one test on a simple test cluster isn't very useful. Let's talk about how we can run more tests on more complex clusters.
+
+## Milestone Two: More Tests!
+
+If you try and simply run all of the conformance tests (the default `E2E_FOCUS` when running Sonobuoy) you will still hit some tests that don’t work well on Windows either by design or due to the early development stage of the Windows tests. As a goal, let’s copy what the upstream sig-windows tests execute and run the tests matching the focus/skip regular expressions:
+
+```
+focus=\[Conformance\]|\[NodeConformance\]|\[sig-windows\]|\[sig-apps\].CronJob|\[sig-api-machinery\].ResourceQuota|\[sig-scheduling\].SchedulerPreemption|\[sig-autoscaling\].\[Feature:HPA\]  
+
+skip=\[LinuxOnly\]|\[Serial\]|GMSA|Guestbook.application.should.create.and.stop.a.working.application
+```
+
+Now you’re targeting the correct set of tests, but you’ll eventually get failures such as:
+
+```
+no matching manifest for windows/amd64 in the manifest list entries
+```
+
+This is because although we have a Sonobuoy Windows image and placed the e2e pod on a Linux node, the e2e tests themselves launch other workloads which use various images. The default images that upstream Kubernetes tests use are not all Windows compatible yet.
+
+To override this we have to tell the tests where to look up test images. This is the same process used to test air-gapped clusters. Typically, users pass this information to Sonobuoy by providing a YAML file which details where the various image repos are and setting a command-line flag. However, Sonobuoy v0.52.0 adds the ability to code ConfigMap information into the plugin definition itself. This means that we will handle coding it into the plugin with no extra work on your part. Luckily for us, [Windows-testing][windowsTesting] provides a matrix of Windows/Kubernetes versions and their respective requirements.
+
+Each one of these will be incorporated into its own plugin so that you only have to point Sonobuoy to the corresponding plugin and the configuration is already correct.
+
+## Milestone Three: Mixed-Node Clusters
+
+Now let’s use a more realistic cluster with both Linux and Windows nodes for workloads. 
+
+First, consider running a Windows specific test case: if it lands on the Linux node we will either get a skip or a failure (depending on the test itself), effectively getting false negatives in our signal. We can also get false positives if a test _would_ fail on a Windows node, but passes on a Linux node so we don’t discover the failure.
+
+As a result, per the windows-testing instructions, it is advisable to taint all of the Linux nodes with a NoSchedule taint to avoid test pods getting placed there. After all the tests complete, we’ll want to remove that NoSchedule taint to return our cluster to its original state.
+
+
+```
+Kubectl taint node <node> sonobuoy:NoSchedule
```

Comment:
Should be lowercase

Reply:
*shakes fist at google docs* Thanks for the catch